File: app/src/main/java/com/zzzmode/appopsx/ui/main/MainListAdapter.java
package com.zzzmode.appopsx.ui.main;

import android.content.Intent;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.zzzmode.appopsx.R;
import com.zzzmode.appopsx.ui.analytics.AEvent;
import com.zzzmode.appopsx.ui.analytics.ATracker;
import com.zzzmode.appopsx.ui.model.AppInfo;
import com.zzzmode.appopsx.ui.permission.AppPermissionActivity;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by zl on 2016/11/18.
 */

class MainListAdapter extends RecyclerView.Adapter<AppItemViewHolder> implements
    View.OnClickListener {

  protected List<AppInfo> appInfos = new ArrayList<>();

  void addItem(AppInfo info) {
    appInfos.add(info);
    notifyItemInserted(appInfos.size() - 1);
  }

  void showItems(List<AppInfo> infos) {
    appInfos.clear();
    if (infos != null) {
      appInfos.addAll(infos);
    }
    notifyDataSetChanged();
  }

  List<AppInfo> getAppInfos() {
    return appInfos;
  }

  @Override
  public AppItemViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    return new AppItemViewHolder(
        LayoutInflater.from(parent.getContext()).inflate(R.layout.item_main_app, parent, false));
  }

  @Override
  public void onBindViewHolder(AppItemViewHolder holder, int position) {
    AppInfo appInfo = appInfos.get(position);
    holder.bindData(appInfo);

    holder.tvName.setText(processText(appInfo.appName));
    holder.itemView.setTag(appInfo);
    holder.itemView.setOnClickListener(this);

  }

  protected CharSequence processText(String name) {
    return name;
  }

  @Override
  public int getItemCount() {
    return appInfos.size();
  }

  @Override
  public void onClick(View v) {
    if (v.getTag() instanceof AppInfo) {

      Intent intent = new Intent(v.getContext(), AppPermissionActivity.class);
      intent.putExtra(AppPermissionActivity.EXTRA_APP, ((AppInfo) v.getTag()));
      v.getContext().startActivity(intent);
      ATracker.send(getAEventId());
    }
  }


  protected String getAEventId() {
    return AEvent.C_APP;
  }

}

File: app/src/main/kotlin/com/zzzmode/appopsx/ui/main/MainListAdapter.kt
package com.zzzmode.appopsx.ui.main

import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.zzzmode.appopsx.R
import com.zzzmode.appopsx.ui.analytics.AEvent
import com.zzzmode.appopsx.ui.analytics.ATracker
import com.zzzmode.appopsx.ui.model.AppInfo
import com.zzzmode.appopsx.ui.permission.AppPermissionActivity
import java.util.ArrayList

/**
 * Created by zl on 2016/11/18.
 */

internal open class MainListAdapter : RecyclerView.Adapter<AppItemViewHolder>(), View.OnClickListener {

    var appInfos: MutableList<AppInfo> = ArrayList()


    protected open val aEventId: String
        get() = AEvent.C_APP

    fun addItem(info: AppInfo) {
        appInfos.add(info)
        notifyItemInserted(appInfos.size - 1)
    }

    fun showItems(infos: List<AppInfo>?) {
        appInfos.clear()
        if (infos != null) {
            appInfos.addAll(infos)
        }
        notifyDataSetChanged()
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AppItemViewHolder {
        return AppItemViewHolder(
                LayoutInflater.from(parent.context).inflate(R.layout.item_main_app, parent, false))
    }

    override fun onBindViewHolder(holder: AppItemViewHolder, position: Int) {
        val appInfo = appInfos[position]
        holder.bindData(appInfo)

        holder.tvName.text = processText(appInfo.appName)
        holder.itemView.tag = appInfo
        holder.itemView.setOnClickListener(this)

    }

    protected open fun processText(name: String?): CharSequence? {
        return name
    }

    override fun getItemCount(): Int {
        return appInfos.size
    }

    override fun onClick(v: View) {


        (v.tag as? AppInfo)?.let {
            val intent = Intent(v.context, AppPermissionActivity::class.java)
            intent.putExtra(AppPermissionActivity.EXTRA_APP, it)
            v.context.startActivity(intent)
            ATracker.send(aEventId)
        }
    }

}
